repa/grids/diffusion.cpp

```diff
@@ -391,5 +365,72 @@ void Diffusion::init_new_foreign_cell(local_cell_index_type localcell,
     util::push_back_unique(borderCellsNeighbors[localcell], owner);
 }
 
+void Diffusion::command(std::string s)
+{
+    std::smatch m;
+
+    static const std::regex profit_re(
+        "(set) (profit) (pass) (through) (([[:digit:]]*[.])?[[:digit:]]+)");
+    if (std::regex_match(s, m, profit_re)) {
+        double ppt = std::stod(m[5].str().c_str(), NULL);
+        if (comm_cart.rank() == 0)
+            std::cout << "Setting profit pass through = " << ppt << std::endl;
+        profit_percentage_pass_through = ppt;
+        return;
+    }
+
+    static const std::regex iter_re("(set) (flow_count) ([[:digit:]]+)");
+    if (std::regex_match(s, m, iter_re)) {
+        uint32_t flow_count = std::stoul(m[3].str().c_str(), NULL);
+        if (comm_cart.rank() == 0)
+            std::cout << "Setting flow_count = " << flow_count << std::endl;
+        try {
+            DIFFUSION_MAYBE_SET_NFLOW_ITER(flow_calc.get(), flow_count);
+        }
+        catch (...) {
+            std::cerr << "Cannot set nflow iter." << std::endl;
+        }
+        return;
+    }
+
+    static const std::regex beta_re(
+        "(set) (beta) (([[:digit:]]*[.])?[[:digit:]]+)");
+    if (std::regex_match(s, m, beta_re)) {
+        double beta_value = std::stod(m[3].str().c_str(), NULL);
+        if (comm_cart.rank() == 0)
+            std::cout << "Setting beta = " << beta_value << std::endl;
+        try {
+            DIFFUSION_MAYBE_SET_BETA(flow_calc.get(), beta_value);
+        }
+        catch (...) {
+            std::cerr << "Cannot set nflow iter." << std::endl;
+        }
+        return;
+    }
+
+    static const std::regex flow_re(
+        "(set) (flow) (willebeek|schornbaum|so|sof)");
+    if (std::regex_match(s, m, flow_re)) {
+        const std::string &impl = m[3].str();
+        if (comm_cart.rank() == 0)
+            std::cout << "Setting implementation to: " << impl << std::endl;
+        if (impl == "willebeek")
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::WILLEBEEK);
+        else if (impl == "schornbaum")
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::SCHORN);
+        else if (impl == "so")
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::SO);
+        else if (impl == "sof")
+            flow_calc = diff_variants::create_flow_calc(
+                diff_variants::FlowCalcKind::SOF);
```

Comment:
Could you use a `std::map` or something like that for this? (Actually, I'd like to use a plain array and designated initializers, however, this feature is not available in C++14).

Reply:
Additional point (please consiver before implementing the map): Do we require that this instance is a PSDiffusion for any of these variants? If so, how do we enforce this?

Reply:
I have just written a patch that addresses this problem, just need to test it. I'll do it in the morning and then commit if everything works.

Reply:
Updated in 337159c.
I have changed it to use inheritance, as I believe that `Diffusion` does not need to know about `so` and `sof`. So I also moved the part with the beta setting to `PSDiffusion`. I also changed it so that the `Setting flow_count = 15` output is only generated when it really works. Also the errors are only displayed on rank 0, which will clean up the output.

Then I changed the macros with the dynamic_cast (DIFFUSION_MAYBE_SET_NFLOW_ITER and DIFFUSION_MAYBE_SET_BETA), because the try catch did not catch errors. This is because an incorrect cast returns a null pointer, which causes a segfault to be thrown because it tries to call a method on the null pointer. I don't know if other compilers would have caught it anyway, gcc-9.3 doesn't do it and with the new version it should be more robust.

Reply:
You're absolutely right, `dynamic_cast` only throws when casting references. For pointers the result is a nullptr if the cast fails.